staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -31,8 +31,69 @@ type DelayingInterface interface {
+	// into the 'active' queue (or dropped) before returning.  Useful for clients/tests that need to
+	// inspect the state of the queue following the add.
+	Synchronous bool
+}
+
+// AddVsWaitingBehaviour defines different choices of how immediate adds interact with waiting items.
+type AddVsWaitingBehaviour int
+
+const (
+	// PreemptWaiting, the default, drops a waiting item when the same item is queued to the 'active' queue.
+	PreemptWaiting AddVsWaitingBehaviour = iota
+	// DropIfWaiting, drops and incoming add when the same item is already in the waiting queue.
+	// This helps us respect rate-limiters and back-offs because we can choose to prioritise the waiting item
+	// over new adds or retries.
+	DropIfWaiting
+	// IgnoreWaiting, will ignore items in the waiting queue and immediately add the item without checking the waiting
+	// queue. Adds made with this option will be slightly faster because they by-pass the waiting check but will result
+	// in superfluous re-executions of the same item (see https://github.com/kubernetes/kubernetes/issues/112325)
+	IgnoreWaiting
+)
+
+// AddAfterOptions toggle implementation specific options for use with calls to AddAfterWithOptions().
+type AddAfterOptions struct {
+	// Existing specifies how you want the waitingLoop to insert items that are already waiting.
+	Existing AlreadyWaitingBehaviour
+	// IgnoreActive, when true, will add a delayed item to the 'waiting' queue when the same item is in the 'active' queue.
+	// The default behaviour is to drop a request to queue a delayed item when the same item is actively queued, because
+	// the actively queued item will soon execute and could fail or succeed, either outcome could render the delayed execution
+	// unnecessary (not recommended - included for completeness and obscure use cases)
+	IgnoreActive bool
+	// Synchronous, when true, instructs AddAfterWithOptions to wait until the item has been proceesed
+	// into either the 'active' or 'waiting' queue (or dropped) before returning.  Useful for clients/tests that need to
+	// inspect the state of the queue following the add.
+	Synchronous bool
+}
+
+// AlreadyWaitingBehaviour defines different choices of how to handle a delayed add when there is already the same waiting 'item'.
+type AlreadyWaitingBehaviour int
+
+const (
+	// TakeShorter will modify the waiting item's wait time only if the incoming item has as shorter delay duration.
+	TakeShorter AlreadyWaitingBehaviour = iota
+	// TakeLonger will modify the waiting item's wait time only if the incoming item has a longer delay duration.
+	TakeLonger
+	// TakeIncoming will always adjust the delay of the waiting 'item' with the delay of the incoming 'item' unless they are equal.
+	TakeIncoming
+	// TakeExisting will never modify the waiting item's wait time.
+	// This is functionally the same as requesting DropIfWaiting on an AddWithOptions() call.
+	TakeExisting
+)
```

Comment:
Could we also consider canceling/forgetting items functionality (https://github.com/kubernetes/client-go/issues/131)?

As mentioned we are are using the delaying queue as a reminder in the controllers for availability check. We schedule the reminder when pod becomes `Ready` It would be great if we could cancel the reminder when the pod losts its readiness or gets deleted.

This feature request is not a must, but it would help the controller do less processing. Example of a usage https://github.com/kubernetes/kubernetes/blob/28e99532f83eeb2d69be56a0539951c687492ffa/pkg/controller/replicaset/replica_set.go#L491 from https://github.com/kubernetes/kubernetes/pull/113605.

I was unaware of this PR and opened mine which implements the forget https://github.com/kubernetes/kubernetes/pull/113607. I think it would be nice to have this as a part of this PR's API.

It is probably not a must at the moment, but wanted to bring awareness to this, and maybe we could add this in a followup?

Reply:
This is why I implemented the Add() pre-empting the waiting item in the first place.  If you are going to queue an immediate processing of 'x' it does not make sense to keep the waiting 'x'.  Either the processing of 'x' is successful in which case the later check is superfluous and fine to be dropped, OR the processing fails (or it not in the desired state) and so a re-check is scheduled with an appropriate delay (i.e. we just checked it, we should back off before we testing/check/process it again).

Reply:
Ok, that makes in most scenarios. But it is not possible to cancel the processing altogether and to clean up the active/waiting queue. I agree, reconcilers should always proceed correctly regardless if the sync is needed or not, so it is not an issue of correctness, but just an optimization to do less processing.

Reply:
Hi Filip

I've added a ForgetDelayed() method along the same lines as your PR - does it look ok?
@atiratree

Reply:
yup, thanks for adding this in! 

I guess the name could be up for a discussion, if somebody has a better one?